### PR TITLE
EOS-18115: CORTX-HA : User interface for cluster stop CLI

### PR DESCRIPTION
**Pull request by Madhura-08**
- Add command line parsing
- Validate arguments

Signed-off-by: Madhura Mande <madhura.mande@seagate.com>

## Problem Statement
<pre>
  <code>
  EOS-18115: CORTX-HA : User interface for cluster stop CLI  </code>
</pre>
## Unit testing on RPM done
<pre>
  <code>
  Yes
  </code>
</pre>
## Problem Description
<pre>
  <code>
    Implement "cortx cluster stop" CLI for HA   </code>
</pre>
## Solution
<pre>
  <code>
    Base framework is already there. Verify cluster add option. . If node is there. Control will go to cluster node add module and eventually to cluster manager
  </code>
</pre>
## Unit Test Cases
<pre>
  <code>
    Execute CLI using 
    cortx cluster stop <--server | --all>. --server is the default argument.
  </code>
</pre>

### CODE REVIEW

File: ha/cli/exec/clusterExecutor.py [hunk 1]
```diff
@@ -134,11 +134,51 @@ def execute(self) -> None:
         Log.info("Cluster started successfully")
 
 class ClusterStopExecutor(CommandExecutor):
+
+    def __init__(self):
+        '''Init method'''
+        super(ClusterStopExecutor, self).__init__()
+        self._pcs_cluster_controller = PcsClusterController()
```
Comment (ajaysrivas) [inline ha/cli/exec/clusterExecutor.py:141]: Should not get PcsClusterController directly. Please change if it was done in earlier PR too.

File: ha/cli/exec/clusterExecutor.py [hunk 2]
```diff
@@ -134,11 +134,51 @@ def execute(self) -> None:
...
+        return False
 
     def execute(self) -> None:
-        raise HAUnimplemented("This operation is not implemented.")
+        '''
+           Execute CLI request by passing it to ClusterManager and
+           also displays an output
+        '''
+        stop_cluster_message = None
+        # stop_cluster_message = self._pcs_cluster_controller.stop()
```
Comment (ajaysrivas) [inline ha/cli/exec/clusterExecutor.py:177]: Call to controller is commented.

**Review by ajaysrivas — CHANGES_REQUESTED** [review]

File: ha/cli/exec/clusterExecutor.py (hunk 1 above)
Comment (ajay-paratmandali) [reply ha/cli/exec/clusterExecutor.py:141]: Example:
----------------------
self._cluster_manager = CortxClusterManager()

for interface check
https://github.com/Seagate/cortx-ha/blob/main/ha/test/integration/cluster_manager/test_cm.py

Comment (Madhura-08) [reply ha/cli/exec/clusterExecutor.py]: Done.

File: ha/cli/exec/clusterExecutor.py (hunk 2 above)
Comment (Madhura-08) [reply ha/cli/exec/clusterExecutor.py]: Now, its uncommented.

File: ha/cli/exec/clusterExecutor.py [hunk 3]
```diff
@@ -66,22 +66,60 @@ def execute(self) -> None:
...
+        '''
+        parser = argparse.ArgumentParser(description="cluster stop")
+        parser.add_argument("cluster", help="Module")
+        parser.add_argument("stop", help="action to be performed")
+        group = parser.add_mutually_exclusive_group()
+        group.add_argument('--all', action='store_true', \
+                            help='Server and storage stop')
+        group.add_argument('--server', action='store_false', \
+                            help='server stop')
+        parser.add_argument('--json', help='Required output format', action='store_true')
```
Comment (ajay-paratmandali) [inline ha/cli/exec/clusterExecutor.py:99]: store_true and store_false what is diff ?

Comment (indrajitzagade) [reply ha/cli/exec/clusterExecutor.py:99]: yeah, we have different options in start I think. Lets understand these.

File: ha/cli/exec/clusterExecutor.py [hunk 4]
```diff
@@ -66,22 +66,60 @@ def execute(self) -> None:
...
+        return False
 
     def execute(self) -> None:
-        raise HAUnimplemented("This operation is not implemented.")
+        '''
+           Execute CLI request by passing it to ClusterManager and
+           also displays an output
+        '''
+        stop_cluster_message = None
+        stop_cluster_message = self.cluster_manager.cluster_controller.stop()
```
Comment (indrajitzagade) [inline ha/cli/exec/clusterExecutor.py:118]: Can we add TODO for storage similar to Start?

File: ha/cli/exec/clusterExecutor.py (hunk 3 above)
Comment (Madhura-08) [reply ha/cli/exec/clusterExecutor.py:99]: store_false will bydefault results to True and store_true will be True if we give that argument. 
Ex: cortx cluster stop
all: False, server: True
cortx cluster stop --all
all: True, server: True
So, default server will be True which is required.

File: ha/cli/exec/clusterExecutor.py (hunk 4 above)
Comment (Madhura-08) [reply ha/cli/exec/clusterExecutor.py:124]: Done.

Comment (indrajitzagade) [reply ha/cli/exec/clusterExecutor.py:124]: ok

File: ha/cli/exec/clusterExecutor.py [hunk 5]
```diff
@@ -66,22 +66,67 @@ def execute(self) -> None:
...
+           Execute CLI request by passing it to ClusterManager and
+           also displays an output
+        '''
+        Log.info("Executing cluster stop")
+        stop_cluster_message = None
+        if self._args.all:
+            Log.info("Executing storageset stop")
+            # TODO: Perform storageset stop
+
+        # TODO: pass the argument parsed from command line. server or all
```
Comment (indrajitzagade) [inline ha/cli/exec/clusterExecutor.py:123]: Is this TODO comment needed now?

File: ha/cli/exec/clusterExecutor.py (hunk 3 above)
Comment (indrajitzagade) [reply ha/cli/exec/clusterExecutor.py:99]: ok

File: ha/cli/exec/clusterExecutor.py (hunk 5 above)
Comment (Madhura-08) [reply ha/cli/exec/clusterExecutor.py]: Hmm..Its not required. Removed.

**Review by indrajitzagade — APPROVED** [review]

File: ha/cli/exec/clusterExecutor.py (hunk 3 above)
Comment (ajay-paratmandali) [reply ha/cli/exec/clusterExecutor.py:99]: Looks good

**Review by ajay-paratmandali — APPROVED** [review]

**Review by azheregelya — APPROVED** [review]

**Review by indrajitzagade — APPROVED** [review]

**Pull request merged**